packages/grid/x-data-grid/src/hooks/features/columns/useGridColumnSpanning.ts

```diff
@@ -0,0 +1,142 @@
+import React from 'react';
+import { useGridApiMethod } from '../../utils/useGridApiMethod';
+import { GridColumnIndex, GridCellColSpanInfo } from '../../../models/gridColumnSpanning';
+import { GridRowId } from '../../../models/gridRows';
+import { GridColumnSpanning } from '../../../models/api/gridColumnSpanning';
+import { useGridApiEventHandler } from '../../utils/useGridApiEventHandler';
+import { GridEvents } from '../../../models/events/gridEvents';
+import { GridCellParams } from '../../../models/params/gridCellParams';
+import { GridStateColDef } from '../../../models/colDef/gridColDef';
+import { GridApiCommunity } from '../../../models/api/gridApiCommunity';
+
+/**
+ * @requires useGridColumns (method, event)
+ * @requires useGridParamsApi (method)
+ */
+export const useGridColumnSpanning = (apiRef: React.MutableRefObject<GridApiCommunity>) => {
+  const lookup = React.useRef<Record<GridRowId, Record<GridColumnIndex, GridCellColSpanInfo>>>({});
+
+  const setCellColSpanInfo = React.useCallback(
+    (rowId: GridRowId, columnIndex: GridColumnIndex, cellColSpanInfo: GridCellColSpanInfo) => {
+      const sizes = lookup.current;
+      if (!sizes[rowId]) {
+        sizes[rowId] = {};
+      }
+
+      sizes[rowId][columnIndex] = cellColSpanInfo;
+    },
+    [],
+  );
+
+  const getCellColSpanInfo = React.useCallback<GridColumnSpanning['unstable_getCellColSpanInfo']>(
+    (rowId, columnIndex) => {
+      return lookup.current[rowId]?.[columnIndex];
+    },
+    [],
+  );
+
+  // Calculate `colSpan` for the cell.
+  const calculateCellColSpan = React.useCallback(
+    (params: {
+      columnIndex: number;
+      rowId: GridRowId;
+      cellParams: GridCellParams;
+      renderedColumns: GridStateColDef[];
+    }) => {
+      const { columnIndex, rowId, cellParams, renderedColumns } = params;
+      const visibleColumns = apiRef.current.getVisibleColumns();
+      const columnsLength = visibleColumns.length;
+      const column = visibleColumns[columnIndex];
+
+      let width = column.computedWidth;
+
+      let colSpan =
+        typeof column.colSpan === 'function' ? column.colSpan(cellParams) : column.colSpan;
+
+      if (typeof colSpan === 'undefined') {
+        colSpan = 1;
+      }
+
+      // Attributes used by `useGridColumnResize` to update column width during resizing.
+      // This makes resizing smooth even for cells with colspan > 1.
+      const dataColSpanAttributes: Record<string, string> = {};
+
+      if (colSpan > 1) {
```

Comment:
To improve readability, if we know that `colSpan==1` we can set the default values and return immediately here.